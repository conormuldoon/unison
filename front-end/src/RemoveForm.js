
import React, { useState } from 'react';
import {API} from './Constant';
import {removePostObject} from './Util';


export default function RemoveForm(props) {


  const [uname,setUName]=useState('');
  const [pword,setPWord]=useState('');


  function handleUN(event){
    setUName(event.target.value);
  }

  function handlePassw(event){
    setPWord(event.target.value);
  }

  function handleSubmit(event){

    removeRequest();

    event.preventDefault();

   }

   async function removeRequest(){

    const response=await fetch(API+'/deleteLocation',
      removePostObject(props.location,uname,pword));

    if(response.ok){
      const resVal=await response.json();

      if(resVal===0){
        alert('Location does not exist');

      }else if(resVal===1){

        alert(props.location+' removed');
        clear();
        props.obtainData();

      }else if (resVal===3){
        alert('You do not have permission to delete this location');
      }else{
        alert('Incorrect user name or password');
      }
    }else{
      alert('Problem connecting');
    }

  }

  function clear(){
    setUName('');
    setPWord('');
    props.hideDisplay();
  }



    let text='Remove Location';
    if(props.display){
      text='Hide';
    }

    return (

      <div>

          <button data-testid='rf-button' onClick={props.toggleDisplay}>{text}</button>

          {props.display&&<form  id='locForm' onSubmit={handleSubmit}>
              <label>
                Username: <input type="text" name="un" onChange={handleUN}  value={uname}  required />
              </label>
              <br/>
              <br/>
              <label>
                Password: <input type="password" name="passwd" onChange={handlePassw}  value={pword} required />
              </label>
              <br/>
              <br/>
              <div>
                  <input  type="submit" value={"Remove "+props.location} />
              </div>
            </form>}
        </div>


    );

}
