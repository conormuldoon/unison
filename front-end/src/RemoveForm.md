RemoveForm example:

```js
import {useState} from 'react';
const [display,setDisplay] = useState(false);
function toggleDisplay(){
    if(display)setDisplay(false);
    else setDisplay(true);
}

<div style={{paddingLeft:200}}>
    <RemoveForm display={display} toggleDisplay ={toggleDisplay} location="Oslo" />
</div>
```