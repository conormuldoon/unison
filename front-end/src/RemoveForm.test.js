import React from 'react';
import ReactDOM from 'react-dom';
import RemoveForm from './RemoveForm';
import { render, cleanup, fireEvent,waitForElement,rerender } from "react-testing-library";

import "@testing-library/jest-dom/extend-expect";
import fetchMock from 'fetch-mock';


it('renders without crashing', async () => {

  const div = document.createElement('div');

  ReactDOM.render(<RemoveForm location='UCD'
    obtainData={()=>{}} hideDisplay={()=>{}} toggleDisplay={()=>{}} display={true}/>, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('mathes snapshot', () => {


  const {container} = render(<RemoveForm display={true} location='UCD' obtainData={()=>{}} hideDisplay={()=>{}} toggleDisplay={()=>{}} display={true}/>);

  expect(container).toMatchSnapshot();

});


it('handles remove location', async (done) =>{
  fetchMock.post('end:/deleteLocation', '1');
  const location='UCD';

  jest.spyOn(window, 'alert').mockImplementation(() => {});
  const hideDisplay=jest.fn();
  
  const obtainData=() =>{
    expect(hideDisplay).toHaveBeenCalledTimes(1);
    expect(alert).toBeCalledWith(location+' removed');
    done();
  };

  const component=<RemoveForm location={location} display={true} obtainData={obtainData} hideDisplay={hideDisplay} toggleDisplay={()=>{}} />;
  const {getByText} = render(component);

  fireEvent.click(getByText('Remove ' + location));

  fetchMock.restore();


});


const testRemove = async(done,retVal,message) =>{
  fetchMock.post('end:/deleteLocation', retVal);
  const location='UCD';

  jest.spyOn(window, 'alert').mockImplementation((message) => {
    expect(message).toEqual(message);
    done();
  });


  const component=<RemoveForm location={location} display={true} obtainData={()=>{}} toggleDisplay={()=>{}} hideDisplay={()=>{}} />;
  const {getByText} = render(component);

  fireEvent.click(getByText('Remove ' + location));

  fetchMock.restore();
}

it('displays that the location is not present', async (done) =>{
  testRemove(done,'0','Location does not exist');

});

it("displays don't have permssion to remove location", async (done) =>{
  testRemove(done,'3','You do not have permission to delete this location');

});

it('Invalid login details', async (done) =>{
  testRemove(done,'2','Incorrect user name or password');

});
