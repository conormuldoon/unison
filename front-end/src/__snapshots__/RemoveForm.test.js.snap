// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mathes snapshot 1`] = `
<div>
  <div>
    <button
      data-testid="rf-button"
    >
      Hide
    </button>
    <form
      id="locForm"
    >
      <label>
        Username: 
        <input
          name="un"
          required=""
          type="text"
          value=""
        />
      </label>
      <br />
      <br />
      <label>
        Password: 
        <input
          name="passwd"
          required=""
          type="password"
          value=""
        />
      </label>
      <br />
      <br />
      <div>
        <input
          type="submit"
          value="Remove undefined"
        />
      </div>
    </form>
  </div>
</div>
`;
